benchmarks/views/index.py

```diff
@@ -30,13 +32,76 @@
 color_suffix = '_color'
 color_None = '#e0e1e2'
 
-
+# Move declaration of the namedtuples to the top of the file to ensure they are available in the cache decorator.
+# Previously inside collect_models function which prevented them from being pickled properly during caching.
+ModelRow = namedtuple('ModelRow', [
+    'id', 'name', 'reference_identifier', 'reference_link',
+    'user', 'public', 'competition', 'domain',
+    'rank', 'scores', 'build_status', 'submitter', 
+    'submission_id', 'jenkins_id', 'timestamp'
+])
+
+ScoreDisplay = namedtuple('ScoreDisplay', [
+    'benchmark', 'versioned_benchmark_identifier',
+    'score_raw', 'score_ceiled', 'error', 'color', 
+    'comment', 'is_complete'
+])
+
+def cache_get_context(timeout=24 * 60 * 60):  # 24 hour cache by default
+    '''
+    Decorator that caches results of get_context function for faster loading of leaderboard view and model card view.
+    Two-level caching:
+        - Global cache for public data
+        - User-specific cache for non-public data
+    Args:
+        timeout (int): Cache timeout in seconds. Defaults to 24 hours.
+    '''
+    def decorator(func):  # Take function to be decorated (i.e., get_context)
+        @wraps(func)  # Preserving original function's metadata attributes
+        def wrapper(user=None, domain="vision", benchmark_filter=None, model_filter=None, show_public=False):
+            # Determine which type of cache key to use
+            if show_public and not user:
+                # (CASE 1: Public data) Create unique key for public data cache
+                key_parts = ['global_context', domain, 'public']
+            elif user:
+                # (CASE 2: User data) Create unique key for user-specific cache
+                key_parts = ['user_context', domain, str(user.id), str(show_public)]
+            else:
+                # (CASE 3: No caching) Neither public nor user-specific
+                return func(user=user, domain=domain, benchmark_filter=benchmark_filter, 
+                          model_filter=model_filter, show_public=show_public)
+            
+            # Generate MD5 hash (to avoid invalid characters, too many characters, etc.). Not necessary but good practice.
+            cache_key = hashlib.md5('_'.join(key_parts).encode()).hexdigest()
```

Comment:
nice - I would probably bump to sha256 vs 128-bit md5 for better collision resistance (and this will keep it inline with other hashes in the Brain-Score ecosystem that use sha256):

`cache_key = hashlib.sha256('_'.join(key_parts).encode()).hexdigest()`

Reply:
Changed!